clippy_lints/src/extern_without_abi.rs

```diff
@@ -0,0 +1,61 @@
+use clippy_utils::diagnostics::span_lint;
+use rustc_ast::visit::FnKind;
+use rustc_ast::{Extern, FnHeader, FnSig, ForeignMod, Item, ItemKind, NodeId};
+use rustc_lint::{EarlyContext, EarlyLintPass};
+use rustc_session::declare_lint_pass;
+use rustc_span::Span;
+
+const LINT_MSG: &str = "`extern` missing explicit ABI";
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Checks for usage of `extern` without an explicit ABI.
+    ///
+    /// ### Why is this bad?
+    /// Explicitly declaring the ABI improves readability.
+    //
+    /// ### Example
+    /// ```no_run
+    /// extern fn foo() {}
+    ///
+    /// extern {
+    ///     fn bar();
+    /// }
+    /// ```
+    /// Use instead:
+    /// ```no_run
+    /// extern "C" fn foo() {}
+    ///
+    /// extern "C" {
+    ///     fn bar();
+    /// }
+    /// ```
+    #[clippy::version = "1.83.0"]
+    pub EXTERN_WITHOUT_ABI,
+    style,
+    "`extern` missing explicit ABI"
+}
+
+declare_lint_pass!(ExternWithoutAbi => [EXTERN_WITHOUT_ABI]);
+
+impl EarlyLintPass for ExternWithoutAbi {
```

Comment:
One issue with the `EarlyLintPass` is that you can't easily ignore procedural macros currently, which I think we should. But the HIR strips whether it's implicitly C or not iirc. In this case checking `snippet == "extern"` should work fine.

Reply:
I moved it to `LateLintPass` and test if `extern` is followed by `fn` or `{`, which would mean the ABI was not specified.